c10/core/DeviceArray.h

```diff
@@ -0,0 +1,21 @@
+#include <c10/core/Allocator.h>
+
+namespace c10 {
+
+template <typename T>
+class DeviceArray {
+ public:
+  DeviceArray(c10::Allocator& allocator, size_t size)
+      : data_ptr_(allocator.allocate(size * sizeof(T))) {
+    static_assert(std::is_pod<T>::value, "T must be 'plain old data'");
+  }
```

Comment:
>check alignment so that get() is ok below?

I don't think that can be done statically as we don't know the allocator's alignment guarantees. (CUDA provides higher alignment by default).

>do we need to construct and destroy the elements with placement new even if they are POD?

I think POD types qualify as [implicit-lifetime types](https://en.cppreference.com/w/cpp/language/lifetime#Implicit-lifetime_types) which means its lifetime begins at allocation even if no constructor is ever called. Also, object lifetime unconditionally ends at deallocation so calling the destructor is only required if it's non-trivial but `is_pod` guarantees it is trivial.

Reply:
> > check alignment so that get() is ok below?
> 
> I don't think that can be done statically as we don't know the allocator's alignment guarantees. (CUDA provides higher alignment by default).
I mean does it make sense to `assert()` that the alignment is suitable. 
> 
> > do we need to construct and destroy the elements with placement new even if they are POD?
> 
> I think POD types qualify as [implicit-lifetime types](https://en.cppreference.com/w/cpp/language/lifetime#Implicit-lifetime_types) which means its lifetime begins at allocation even if no constructor is ever called. Also, object lifetime unconditionally ends at deallocation so calling the destructor is only required if it's non-trivial but `is_pod` guarantees it is trivial.
Isn't this guarantee only true for specific allocation functions? Is the CUDA allocator included in this list?

Can the requirement of this be loosened to `std::is_trivial`?

Reply:
>Isn't this guarantee only true for specific allocation functions? 

You can read the [cppreference text here](https://en.cppreference.com/w/cpp/language/object#Object_creation). The allocations functions listed are I believe every standard way to allocate raw memory (`operator new`, `malloc` and arrays of bytes). I believe the intention is to support C-style code where constructors don't exist.

> Is the CUDA allocator included in this list?

CUDA isn't really relevant to the C++ standard here because the C++ compiler can't dereference CUDA memory at all. The pattern of just assigning to CUDA memory is used ubiquitously in CUDA examples so is clearly supported.

> Can the requirement of this be loosened to `std::is_trivial`?

I think that's right. Changed and rebased on `viable/strict`.

Reply:
> > > check alignment so that get() is ok below?
> > 
> > 
> > I don't think that can be done statically as we don't know the allocator's alignment guarantees. (CUDA provides higher alignment by default).
> 
> I mean does it make sense to `assert()` that the alignment is suitable.

I messed up the quoted reply earlier, what do you think of this?
> 
> > > do we need to construct and destroy the elements with placement new even if they are POD?
> > 
> > 
> > I think POD types qualify as [implicit-lifetime types](https://en.cppreference.com/w/cpp/language/lifetime#Implicit-lifetime_types) which means its lifetime begins at allocation even if no constructor is ever called. Also, object lifetime unconditionally ends at deallocation so calling the destructor is only required if it's non-trivial but `is_pod` guarantees it is trivial.
> 
> Isn't this guarantee only true for specific allocation functions? Is the CUDA allocator included in this list?
> 
> Can the requirement of this be loosened to `std::is_trivial`?

Reply:
> > Isn't this guarantee only true for specific allocation functions?
> 
> You can read the [cppreference text here](https://en.cppreference.com/w/cpp/language/object#Object_creation). The allocations functions listed are I believe every standard way to allocate raw memory (`operator new`, `malloc` and arrays of bytes). I believe the intention is to support C-style code where constructors don't exist.
> 
> > Is the CUDA allocator included in this list?
> 
> CUDA isn't really relevant to the C++ standard here because the C++ compiler can't dereference CUDA memory at all. The pattern of just assigning to CUDA memory is used ubiquitously in CUDA examples so is clearly supported.

I'm not super worried about this, but the cppreference page seems to indicate that the implicit lifetimes are limited to the enumerated allocator functions. Is `c10::Allocator` guaranteed to use one of them?
> 
> > Can the requirement of this be loosened to `std::is_trivial`?
> 
> I think that's right. Changed and rebased on `viable/strict`.

Reply:
>I mean does it make sense to `assert()` that the alignment is suitable.
Okay, I've added an assert.

> I'm not super worried about this, but the cppreference page seems to indicate that the implicit lifetimes are limited to the enumerated allocator functions. Is c10::Allocator guaranteed to use one of them?

The CPU allocator uses `posix_memalign` which isn't part of standard C++ so can't be included in that list. However, it's notionally the same as `malloc` so I don't think it will be an issue.

Reply:
sounds good, thanks!